jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java

```diff
@@ -504,7 +577,11 @@ private boolean tryRelease()
                     return false;
                 int newMultiplexCount = multiplexCount - 1;
                 if (state.compareAndSet(encoded, 0, newMultiplexCount))
+                {
+                    if (pool.weak && newMultiplexCount == 0)
+                        this.released = pooled.get();
```

Comment:
GC can run before this line, and `pooled.get()` returns `null`, so the entry is released, but it's garbage.

Reply:
@sbordet @lorban  How can this happen?  Specifically, how can tryRelease ever be called without a hard reference to the entry?

In the case of a `RetainableByteBuffer`, it is created with:
```java
                buffer = newRetainableByteBuffer(bucket._capacity, direct, retainedBuffer ->
                {
                    BufferUtil.reset(retainedBuffer.getByteBuffer());
                    if (!reservedEntry.release())
                        reservedEntry.remove();
                });
```

So the lambda holds a reference to the `reservedEntry`, which will then not be GC'd until such time as the `RetainableByteBuffer` is not referenced.       So is it possible for a RBB to be released without a reference being held to the entry?

Reply:
OHHHHH!!!!! The week reference is to the pooled object, not the entry.   So an application can hold a strong reference to the entry, but not to the pooled item.  So GC clean of the pooled item and release of the entry are in a race!

I don't think this approach can work.      It could only work if the weak reference is to the same instance as the application calls release on..... but then I don't think we have enough info to do anything with a collected reference, because we will not know which entry it was!